File: weisi/src/main/java/com/cjq/lib/weisi/communicator/tcp/TcpClient.java
package com.cjq.lib.weisi.communicator.tcp;

import android.support.annotation.NonNull;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;

/**
 * Created by CJQ on 2018/3/12.
 */

public class TcpClient extends Tcp {

    private TcpSocket mSocket;

    public void connect(@NonNull final String serverIp, final int serverPort,
                        @NonNull final OnServerConnectListener listener) {
        connect(serverIp, serverPort, listener, 0);
    }

    public void connect(@NonNull final String serverIp, final int serverPort,
                        @NonNull final OnServerConnectListener listener,
                        final int timeout) {
        if (mSocket == null && mState != CONNECTING) {
            mState = CONNECTING;
            new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        Socket socket = new Socket();
                        socket.connect(new InetSocketAddress(serverIp, serverPort), timeout);
                        if (socket.isConnected()) {
                            mSocket = new TcpSocket(socket);
                            mState = CONNECTED;
                        }
                    } catch (IOException e) {
                        mSocket = null;
                        mState = UNCONNECTED;
                        e.printStackTrace();
                    } finally {
                        listener.onServerConnect(mState, mSocket);
                    }
                }
            }).start();
        } else {
            listener.onServerConnect(mState, mSocket);
        }
    }

    public TcpSocket getSocket() {
        return mSocket;
    }

    public void shutdown() {
        if (mSocket != null) {
            try {
                mSocket.close();
                mSocket = null;
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                mState = UNCONNECTED;
            }
        }
    }

    public interface OnServerConnectListener {
        void onServerConnect(@ConnectState int state, TcpSocket socket);
    }
}

File: wsb/src/main/java/com/wsn/lib/wsb/communicator/tcp/TcpClient.kt
package com.wsn.lib.wsb.communicator.tcp

import java.io.IOException
import java.net.InetSocketAddress
import java.net.Socket

import com.wsn.lib.wsb.communicator.tcp.Tcp.ConnectState.CONNECTED
import com.wsn.lib.wsb.communicator.tcp.Tcp.ConnectState.CONNECTING
import com.wsn.lib.wsb.communicator.tcp.Tcp.ConnectState.UNCONNECTED

/**
 * Created by CJQ on 2018/3/12.
 */

class TcpClient : Tcp() {

    lateinit var tcpSocket: TcpSocket
        private set

    @JvmOverloads
    fun connect(serverIp: String, serverPort: Int,
                listener: OnServerConnectListener,
                timeout: Int = 0) {
        if (state == UNCONNECTED) {
            state = CONNECTING
            Thread(Runnable {
                try {
                    val socket = Socket()
                    socket.connect(InetSocketAddress(serverIp, serverPort), timeout)
                    if (socket.isConnected) {
                        tcpSocket = TcpSocket(socket)
                        state = CONNECTED
                    }
                } catch (e: IOException) {
                    shutdown()
                    state = UNCONNECTED
                    e.printStackTrace()
                } finally {
                    notifyListener(listener)
                }
            }).start()
        } else {
            notifyListener(listener)
        }
    }

    private fun notifyListener(listener: OnServerConnectListener) {
        listener.onServerConnect(state, if (state == CONNECTED) {
            tcpSocket
        } else {
            null
        })
    }

    fun shutdown() {
        if (state == CONNECTED) {
            try {
                tcpSocket.close()
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                state = UNCONNECTED
            }
        }
    }

    interface OnServerConnectListener {
        fun onServerConnect(state: Tcp.ConnectState, socket: TcpSocket?)
    }
}
